pkg/cmd/provisioning/nutanix/create_shared_secrets.go

```diff
@@ -57,70 +81,123 @@ func createSharedSecretsCmd() *cobra.Command {
+	if filePath == "" {
+		homePath := GetHomePath()
+		if homePath == "" {
+			return errors.New("The default source credential file path is invalid")
+		}
+		filePath = filepath.Join(homePath, ".nutanix", "credentials")
 	}
-	port := os.Getenv(portKeyEnvVar)
-	if port == "" {
-		return fmt.Errorf("%s environment variable not set", portKeyEnvVar)
+
+	if _, err := os.Stat(filePath); err != nil {
+		return fmt.Errorf("The source credentials file %s does not exist.", filePath)
 	}
-	uname := os.Getenv(usernameKeyEnvVar)
-	if uname == "" {
-		return fmt.Errorf("%s environment variable not set", usernameKeyEnvVar)
+
+	bytes, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return fmt.Errorf("Failed to read the credentials file %s. %v", filePath, err)
 	}
-	pass := os.Getenv(passwordKeyEnvVar)
-	if pass == "" {
-		return fmt.Errorf("%s environment variable not set", passwordKeyEnvVar)
+
+	creds := &NutanixCredentials{}
+	if err = yaml.Unmarshal(bytes, creds); err != nil {
+		return fmt.Errorf("Failed to convert the credentials data read from file %s. %v", filePath, err)
 	}
 
-	err := createSecrets(CreateSharedSecretsOpts.CredRequestDir, CreateSharedSecretsOpts.TargetDir, endpoint, port, uname, pass, CreateSharedSecretsOpts.EnableTechPreview)
+	retCreds, err := getCredentialsData(creds)
+	if err != nil {
+		return errors.Wrapf(err, "The credentials data read from file %s is invalid.", filePath)
+	}
+
+	err = createSecrets(CreateSharedSecretsOpts.CredRequestDir, CreateSharedSecretsOpts.TargetDir, retCreds, CreateSharedSecretsOpts.EnableTechPreview)
 	if err != nil {
 		return errors.Wrap(err, "Failed to create credentials secrets")
 	}
+
 	return nil
 }
 
-func createSecrets(credReqDir, targetDir, endpoint, port, username, password string, enableTechPreview bool) error {
+// Retrieve the credentials data
+func getCredentialsData(creds *NutanixCredentials) (*NutanixCredentials, error) {
+	var err error
+	retCreds := &NutanixCredentials{}
+	for _, cred := range creds.Credentials {
+		switch cred.Type {
+		case BasicAuthCredentialType:
+			basicAuthCreds := &BasicAuthCredential{}
+			if err = yaml.Unmarshal(cred.Data.Raw, basicAuthCreds); err != nil {
+				return nil, errors.Wrap(err, "Failed to convert the basic-auth data.")
+			}
+			if basicAuthCreds.PrismCentral.Username == "" || basicAuthCreds.PrismCentral.Password == "" {
+				return nil, errors.New("The Prism Central username and/or password are not set correctly.")
+			}
+			if len(basicAuthCreds.PrismElements) == 0 || basicAuthCreds.PrismElements[0].Name == "" ||
+				basicAuthCreds.PrismElements[0].Username == "" || basicAuthCreds.PrismElements[0].Password == "" {
+				return nil, errors.New("The Prism Element (cluster) username and/or password are not set correctly.")
+			}
+
+			credsJsonBytes, err := json.Marshal(*basicAuthCreds)
+			if err != nil {
+				return nil, errors.Errorf("Failed to convert the basic_auth type credentials object to json. %v", err)
```

Comment:
I believe this function validates if credentials are correct and have the necessary data. If I understand correctly there can be multiple credentials. Do we need to error out if one of the credentials is invalid? I may be missing something but can't we create secrets with only valid credentials?

Reply:
We do allow more than one set of credential data in the input. Currently we only support the "basic_auth" credential type, and we may extend to support other credential types in the future. We expect all the data in the input file is correctly formatted and ccoctl will fail with error if not. This will help users notice and correct any format errors in the input data file.